prometric/src/summary/batching.rs

```diff
@@ -0,0 +1,211 @@
+        // current batch
+
+        // NOTE: `SplitVec` can't be initialized with a requested total capacity directly
+        let mut batch = Batch::new();
+        batch.reserve_maximum_capacity(batch_size);
+
+        Arc::new(batch)
+    }
+
+    /// Wait for the given Arc to have a single owner and obtain the inner value
+    pub(crate) fn wait_for_arc<T>(mut arc: Arc<T>) -> T {
+        loop {
+            match Arc::try_unwrap(arc) {
+                Ok(inner) => return inner,
+                Err(this) => {
+                    arc = this;
+                }
+            }
+
+            std::hint::spin_loop();
+        }
+    }
+
+    /// Commits the current measurements batch to the underlying summary
+    ///
+    /// Will clear current the measurements batch
+    pub fn commit(&self) {
+        // If [`FixedBatch`] had something like `.take()` the [`ArcCell`] would be unnecessary
+        // NOTE: we take the previous batch so new measurements can be added without changing
+        // the set that we are currently committing
+        let measurements = self.measurements.set(Self::new_batch(self.batch_size));
+        let measurements = Self::wait_for_arc(measurements);
+
+        let mut inner = self.inner.write();
+
+        for measure in measurements.into_iter() {
+            inner.observe(measure);
+        }
+    }
+
+    /// Retrieve the inner summary
+    ///
+    /// Will commit the current batch before returning the summary
+    pub fn into_inner(self) -> P {
+        self.commit();
+        self.inner.into_inner()
+    }
+}
+
+impl<P: SummaryProvider> SummaryProvider for BatchedSummary<P> {
+    type Opts = BatchOpts<P::Opts>;
+    type Summary = P::Summary;
+
+    fn new(opts: &Self::Opts) -> Self {
+        let inner = RwLock::new(P::new(&opts.inner));
+        Self {
+            inner,
+            measurements: ArcCell::new(Self::new_batch(opts.batch_size)),
+            batch_size: opts.batch_size,
+        }
+    }
+
+    fn observe(&mut self, val: f64) {
+        self.concurrent_observe(val);
+    }
+
+    fn snapshot(&self) -> Self::Summary {
+        // Forcefully commit the current batch before snapshotting
+        self.commit();
+        self.inner.read().snapshot()
+    }
+}
+
+impl<P: SummaryProvider> ConcurrentSummaryProvider for BatchedSummary<P> {
+    fn concurrent_observe(&self, val: f64) {
+        let measurements = self.measurements.get();
+        measurements.push(val);
+
+        if measurements.len() >= self.batch_size {
+            // forcefully drop the guard before committing
+            // to avoid deadlocks
+            std::mem::drop(measurements);
+
+            // Commit the current batch
+            self.commit()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use crate::{
+        simple::{SimpleSummary, SimpleSummaryOpts},
+        traits::Summary,
+    };
+
+    use super::*;
+
+    #[tokio::test]
```

Comment:
Any reason we're pulling in Tokio here instead of doing it with std::thread?

Reply:
Yeah no I did it out of habit and I thought I’d also try in an async context but yeah I should definitely just use thread